server/channels/app/authentication_test.go

```diff
@@ -55,3 +59,111 @@ func TestParseAuthTokenFromRequest(t *testing.T) {
 	}
 }
+
+func TestCheckPasswordAndAllCriteria(t *testing.T) {
+	th := Setup(t).InitBasic()
+	defer th.TearDown()
+
+	const maxFailedLoginAttempts = 3
+	const concurrentAttempts = maxFailedLoginAttempts + 1
+	th.App.UpdateConfig(func(cfg *model.Config) {
+		*cfg.ServiceSettings.MaximumLoginAttempts = maxFailedLoginAttempts
+		*cfg.ServiceSettings.EnableMultifactorAuthentication = true
+	})
+
+	password := "newpassword1"
+	appErr := th.App.UpdatePassword(th.Context, th.BasicUser, password)
+	require.Nil(t, appErr)
+
+	// setup MFA
+	secret, appErr := th.App.GenerateMfaSecret(th.BasicUser.Id)
+	require.Nil(t, appErr)
+	err := th.Server.Store().User().UpdateMfaActive(th.BasicUser.Id, true)
+	require.NoError(t, err)
+	err = th.Server.Store().User().UpdateMfaSecret(th.BasicUser.Id, secret.Secret)
+	require.NoError(t, err)
+
+	t.Run("should run successfully when attempts are available", func(t *testing.T) {
+		err = th.App.Srv().Store().User().UpdateFailedPasswordAttempts(th.BasicUser.Id, maxFailedLoginAttempts-1)
+		require.NoError(t, err)
+		code := dgoogauth.ComputeCode(secret.Secret, time.Now().UTC().Unix()/30)
+		token := fmt.Sprintf("%06d", code)
+
+		appErr = th.App.CheckPasswordAndAllCriteria(th.Context, th.BasicUser.Id, password, token)
+		require.Nil(t, appErr)
+	})
+
+	t.Run("validate concurrent failed attempts to bypass checks", func(t *testing.T) {
+		testCases := []struct {
+			name          string
+			password      string
+			mfaToken      string
+			expectedErrID string
+		}{
+			{
+				name:          "should not breach max. login attempts when password is wrong",
+				password:      "wrong password",
+				expectedErrID: "api.user.check_user_password.invalid.app_error",
+			},
+			{
+				name:          "should not breach max. login attempts when MFA is wrong",
+				password:      password,
+				mfaToken:      "123456",
+				expectedErrID: "api.user.check_user_mfa.bad_code.app_error",
+			},
+		}
+
+		for _, tc := range testCases {
+			t.Run(tc.name, func(t *testing.T) {
+				// Reset login attempts
+				th.App.Srv()
+				err := th.App.Srv().Store().User().UpdateFailedPasswordAttempts(th.BasicUser.Id, 0)
+				require.NoError(t, err)
+
+				// Capture all concurrent errors
+				appErrs := make([]*model.AppError, concurrentAttempts)
+
+				// Wait to complete the test
+				var completeWG sync.WaitGroup
+				completeWG.Add(concurrentAttempts)
+
+				// Wait to fetch the login attempts
+				var fetchedWG sync.WaitGroup
+				fetchedWG.Add(concurrentAttempts)
+
+				for i := 0; i < concurrentAttempts; i++ {
+					go func(i int) {
+						defer completeWG.Done()
+
+						// Simulate concurrent login attempts for the same user
+						user, _ := th.App.Srv().Store().User().Get(th.Context.Context(), th.BasicUser.Id)
+						fetchedWG.Done()
+
+						// Simulate that all userObjects/failedAttemptsCount have been fetched
+						// before any of the goroutine ever reaches the password/mfa checks
+						fetchedWG.Wait()
```

Comment:
Wondering if it makes much sense to do this artificial dance, given that anyways we are simulating the whole thing. I'd just get the user.Id once outside the for loop, and call `th.App.CheckPasswordAndAllCriteria` with the given userID. It would remove any need for the `fetchedWG`.

Reply:
Technically yes, we can remove this. But this clearly highlights the real scenario that is actually being tested.
If you strongly feel about this, LMK, I'll remove it.

Reply:
I think the complicacy in the test is not worth highlighting this particular scenario when it's not bringing any additional code coverage. But will defer to other reviewers.

Reply:
I think we should remove the `fetchedWG` and leave a detailed comment highlighting what the scenario is (probably at the top of the test so it isn't buried).